dbeaver/dbeaver#36814 Implemented support for editing table and column comments in the GBase8s driver

Wintianis opened the pull request:
### Overview:
This PR introduces the ability to edit table and column comments in the GBase8s driver, addressing a feature gap in the database management capabilities.
### Key Changes:
- Implemented support for editing comments at the table and column levels.
### Related Issue:
Closes dbeaver/dbeaver#36814.
### Testing:
*  Created new columns with comments.
*  Added comments to existing columns.
*  Removed comments from columns.
*  Deleted columns with comments.
*  All test cases have been successfully validated and executed to ensure proper comment operations.
### Scope of Impact:
This change affects only the GBase8s driver and is backward-compatible.
### Checklist:
*  Local tests passed.
*  No breaking changes identified.

Wintianis committed:
dbeaver/dbeaver#36814 Initial commit: Implemented support for editing table and column comments in the GBase8s driver

ShadelessFox commented on plugins/org.jkiss.dbeaver.ext.gbase8s/src/org/jkiss/dbeaver/ext/gbase8s/model/meta/GBase8sMetaModel.java:
```diff
@@ -288,6 +292,64 @@ public JDBCStatement prepareTableLoadStatement(@NotNull JDBCSession session, @No
+                : JDBCUtils.escapeWildCards(session, owner.getSchema().getName());
+        boolean isOracleMode = GBase8sUtils.isOracleSqlMode(owner.getDataSource().getContainer());
+        String ownerPattern = isOracleMode ? schema + "." : catalog + ":";
+        String sql = "SELECT"
```
Consider using text blocks `"""`

ShadelessFox commented on plugins/org.jkiss.dbeaver.ext.gbase8s/src/org/jkiss/dbeaver/ext/gbase8s/edit/GBase8sTableColumnManager.java:
```diff
@@ -35,13 +34,25 @@
+    protected void addObjectCreateActions(DBRProgressMonitor monitor, DBCExecutionContext executionContext,
+            List<DBEPersistAction> actions,
+            SQLObjectEditor<GenericTableColumn, GenericTableBase>.ObjectCreateCommand command,
+            Map<String, Object> options) throws DBException {
```
Formatting, missing annotations
```suggestion
    protected void addObjectCreateActions(
        @NotNull DBRProgressMonitor monitor,
        @NotNull DBCExecutionContext executionContext,
        @NotNull List<DBEPersistAction> actions,
        @NotNull SQLObjectEditor<GenericTableColumn, GenericTableBase>.ObjectCreateCommand command,
        @NotNull Map<String, Object> options
    ) throws DBException {
```

ShadelessFox commented on plugins/org.jkiss.dbeaver.ext.gbase8s/src/org/jkiss/dbeaver/ext/gbase8s/model/meta/GBase8sMetaModel.java:
```diff
@@ -323,12 +385,35 @@ JDBCStatement prepareTSObjectLoadStatement(@NotNull JDBCSession session, @NotNul
-                isOracleMode ? schemaPattern : catalog, tableNamePattern, types).getSourceStatement();
+        String ownerPattern = isOracleMode ? schema + "." : catalog + ":";
+        
+        String sql = "SELECT t.tabid, t.tabname AS TABLE_NAME, t.owner AS "
```
Consider using text blocks `"""`

ShadelessFox commented on plugins/org.jkiss.dbeaver.ext.gbase8s/src/org/jkiss/dbeaver/ext/gbase8s/model/meta/GBase8sMetaModel.java:
```diff
@@ -323,12 +385,35 @@ JDBCStatement prepareTSObjectLoadStatement(@NotNull JDBCSession session, @NotNul
+            }
+        }
+        return dbStat;
+        // The getTables(...) interface returns a result that does not include information about table comments.
+//        return session.getMetaData().getTables(isOracleMode ? catalog : schemaPattern,
+//                isOracleMode ? schemaPattern : catalog, tableNamePattern, types).getSourceStatement();
```
Please remove the commented-out code. This explanation is nice, but rather pointless as custom implementations of meta-models are expected to override the default behavior.

Wintianis committed:
Refactor string concatenation to use text blocks

Wintianis committed:
Fix formatting, add missing annotations, and resolve an issue found during self-testing

E1izabeth commented:
Thanks for your contribution! We passed your pull request to out QA team for testing

Wintianis committed:
Fix the error that occurs when the Connection Properties setting is missing or empty

Wintianis commented:
I have fixed an identified issue. Please review.

ShadelessFox commented on plugins/org.jkiss.dbeaver.ext.gbase8s/src/org/jkiss/dbeaver/ext/gbase8s/GBase8sUtils.java:
```diff
@@ -67,16 +69,20 @@ public static List<String> getSource(DBRProgressMonitor monitor, String sqlState
-        return false;
+        DBPDriver driver = container.getDriver();
+        Map<String, Object> driverProps = driver.getConnectionProperties();
+        return driverProps.values().stream()
+                .filter(Objects::nonNull)
+                .map(Object::toString)
+                .map(String::trim)
+                .anyMatch(value -> "oracle".equalsIgnoreCase(value));
```
Why does it check only values without the key?

Wintianis replied:
I have addressed the issue with the logic where the key was not being considered. The code has been updated, and I would appreciate it if you could review it again.

Commit:
Fixed the issue where the key was not considered in the logic

serge-rider commented on plugins/org.jkiss.dbeaver.ext.gbase8s/src/org/jkiss/dbeaver/ext/gbase8s/GBase8sUtils.java:
```diff
@@ -67,16 +68,21 @@ public static List<String> getSource(DBRProgressMonitor monitor, String sqlState
+        DBPConnectionConfiguration configuration = container.getConnectionConfiguration();
+        Map<String, String> cprops = configuration.getProperties();
+        for (String key : cprops.keySet()) {
+            if ("sqlmode".equalsIgnoreCase(key)) {
```
Please move hardcoded sqlmode and oracle into constants

ShadelessFox commented on plugins/org.jkiss.dbeaver.ext.gbase8s/src/org/jkiss/dbeaver/ext/gbase8s/GBase8sUtils.java:
```diff
@@ -78,11 +77,12 @@ public static boolean isOracleSqlMode(DBPDataSourceContainer container) {
-                .anyMatch(value -> "oracle".equalsIgnoreCase(value));
+        for (String key : driverProps.keySet()) {
+            if ("sqlmode".equalsIgnoreCase(key)) {
+                return "oracle".equalsIgnoreCase(Objects.toString(driverProps.get(key), "").trim());
```
The code above doesn't check for `null` values. Maybe it's better to extract that part to avoid repetition?

Commit:
Added null checks, refactored duplicated code, and replaced hardcoded values with constants

ShadelessFox commented on plugins/org.jkiss.dbeaver.ext.gbase8s/src/org/jkiss/dbeaver/ext/gbase8s/GBase8sUtils.java:
```diff
@@ -67,16 +68,32 @@ public static List<String> getSource(DBRProgressMonitor monitor, String sqlState
+        return isOracleSqlModeFromProperties(driverProps);
+    }
+
+    private static boolean containsSqlModeKey(Map<String, ?> properties) {
+        return properties != null && !properties.isEmpty() && properties.keySet().stream()
+                .anyMatch(key -> key != null && GBase8sConstants.JDBC_SQL_MODE.equalsIgnoreCase(key));
+    }
+
+    private static boolean isOracleSqlModeFromProperties(Map<String, ?> properties) {
+        return properties != null && !properties.isEmpty()
+                && properties.entrySet().stream()
+                        .anyMatch(entry -> entry.getKey() != null
+                                && GBase8sConstants.JDBC_SQL_MODE.equalsIgnoreCase(entry.getKey().toString())
+                                && GBase8sConstants.JDBC_SQL_MODE_ORACLE
+                                        .equalsIgnoreCase(Objects.toString(entry.getValue(), "").trim()));
```
Can you please extract the value of `JDBC_SQL_MODE` and compare it instead of doing this again the second time? Why are you converting the key to a string even though it's already a string? Also, please stop abusing stream APIs.

Wintianis replied:
I got it, I have already fixed it. If there are any further issues, please feel free to contact me.

Commit:
Modify the judgment logic of JDBC_SQL_Mode

ShadelessFox reviewed (dismissed):
Looks good. Thank you for cooperation :)

E1izabeth commented:
Passed to QA team for testing

Commit:
Fix issues with "columns not editable via Metadata Editor" and "comments not included in table DDL"

Wintianis commented:
I have submitted the fix for the bug reported by QA. Please review the code. Apologies for any inconvenience caused. Thank you very much.
@ShadelessFox @E1izabeth @serge-rider

ShadelessFox commented on plugins/org.jkiss.dbeaver.ext.gbase8s/src/org/jkiss/dbeaver/ext/gbase8s/GBase8sConstants.java:
```diff
@@ -26,4 +26,6 @@ public class GBase8sConstants {
     public static final String JDBC_SQL_MODE_ORACLE = "oracle";
     public static final String JDBC_SQL_MODE_GBASE = "gbase";
 
+    public static final String SQL_TABLE_COMMENT = "COMMENT ON TABLE %s IS '%s'";
```
This is a potential place for an SQL injection. Please use `SQLUtils.quoteString` to escape string literals, and `DBUtils#getQuotedIdentifier` to escape object names.

Wintianis replied:
I have fixed it. Please review the code.

Commit:
Fix potential SQL injection vulnerabilities

ShadelessFox commented on plugins/org.jkiss.dbeaver.ext.gbase8s/src/org/jkiss/dbeaver/ext/gbase8s/edit/GBase8sTableManager.java:
```diff
@@ -120,9 +122,10 @@ private DBEPersistAction buildTableCommentAction(GenericTableBase table) {
-            String tableName = column.getTable().getFullyQualifiedName(DBPEvaluationContext.DDL);
-            String columnName = column.getName();
-            String commentSQL = String.format(GBase8sConstants.SQL_COLUMN_COMMENT, tableName, columnName, description);
+            String tableName = DBUtils.getObjectFullName(column.getTable(), DBPEvaluationContext.DDL);
+            String columnName = DBUtils.getQuotedIdentifier(column);
```
If you call `DBUtils.getObjectFullName` on a column, it will return a fully-qualified identifier.

Wintianis replied:
Thank you for your comment.
I agree with your suggestion. The method `DBUtils.getObjectFullName(column.getTable(), DBPEvaluationContext.DDL)` is intentionally used here to ensure that the fully-qualified table name is obtained instead of the column identifier. Please let me know if you have any further suggestions.

labanovichttps commented on plugins/org.jkiss.dbeaver.ext.gbase8s/src/org/jkiss/dbeaver/ext/gbase8s/GBase8sUtils.java:
```diff
@@ -67,16 +66,25 @@ public static List<String> getSource(DBRProgressMonitor monitor, String sqlState
+        }
+        for (Map.Entry<String, ?> entry : properties.entrySet()) {
+            if (GBase8sConstants.JDBC_SQL_MODE.equalsIgnoreCase(entry.getKey())) {
+                return entry.getValue() != null ? entry.getValue().toString().trim() : null;
```
Use **org.jkiss.utils.CommonUtils#trim**

ShadelessFox replied:
It expects a `String`

labanovichttps commented on plugins/org.jkiss.dbeaver.ext.gbase8s/src/org/jkiss/dbeaver/ext/gbase8s/edit/GBase8sTableManager.java:
```diff
@@ -64,6 +69,28 @@ public void renameObject(DBECommandContext commandContext, GenericTableBase obje
 
+    @Override
+    protected void addStructObjectCreateActions(
+            DBRProgressMonitor monitor,
+            DBCExecutionContext executionContext,
+            List<DBEPersistAction> actions,
+            StructCreateCommand command,
+            Map<String, Object> options) throws DBException {
```
Add annotations(`@NotNull, @Nullable`), please

Wintianis replied:
I have added the annotations.

labanovichttps commented on plugins/org.jkiss.dbeaver.ext.gbase8s/src/org/jkiss/dbeaver/ext/gbase8s/model/meta/GBase8sMetaModel.java:
```diff
@@ -288,17 +294,82 @@ public JDBCStatement prepareTableLoadStatement(@NotNull JDBCSession session, @No
+        String schema = owner.getSchema() == null || DBUtils.isVirtualObject(owner.getSchema()) ? null
+                : JDBCUtils.escapeWildCards(session, owner.getSchema().getName());
+        boolean isOracleMode = GBase8sUtils.isOracleSqlMode(owner.getDataSource().getContainer());
+        String ownerPattern = """
+                %s%s
+                """.formatted(isOracleMode ? schema : catalog, isOracleMode ? "." : ":");
```
If it's a simple string, maybe use this code.
` String ownerPattern = "%s%s".formatted(isOracleMode ? schema : catalog, isOracleMode ? "." : ":");`

Wintianis replied:
As requested, I have modified the implementation.

Commit:
Optimize simple string replacement writing

Commit:
Optimize the generation of table & column comment DDL

Wintianis commented:
I have updated part of the comment DDL generation logic and added the `DBPScriptObject.OPTION_INCLUDE_COMMENTS` option button. Meanwhile, I moved the comment DDL generation code to `GBase8sTableManager.addObjectExtraActions` and `GBase8sTableColumnManager.addObjectExtraActions`, which might be a more standardized approach.  

If there is anything inappropriate, please feel free to point it out. Your review is always welcome.

Wintianis committed:
Exclude column comments from the table comment script when adding a table comment via the Metadata Editor

Wintianis commented:
I have adjusted the logic for outputting column comments based on the testing suggestion. Please review the code.

kseniaguzeeva commented:
verified
